Maple2.Server.Game/PacketHandlers/MeretMarketHandler.cs

```diff
@@ -0,0 +1,310 @@
+        MeretMarketSection section = GetMarketSection(packet.ReadByte());
+
+        ICollection<MarketItem> entries = GetItems(session, section, sortBy, 0, gender, job, searchString);
+        int totalItems = entries.Count;
+        entries = TakeLimit(entries, startPage, itemsPerPage);
+
+        session.Send(MeretMarketPacket.LoadItems(entries.ToList(), totalItems, startPage));
+    }
+
+    private static MeretMarketSection GetMarketSection(byte section) {
+        return section switch {
+            0 => MeretMarketSection.All,
+            1 => MeretMarketSection.Premium,
+            2 => MeretMarketSection.RedMeret,
+            3 => MeretMarketSection.Ugc,
+            _ => MeretMarketSection.All,
+        };
+    }
+
+    private ICollection<MarketItem> GetItems(GameSession session, MeretMarketSection section, MeretMarketSort sortBy, int tabId, GenderFilterFlag genderFilter, JobFilterFlag jobFilter, string searchString) {
+        using GameStorage.Request db = session.GameStorage.Context();
+        if (tabId == 0) {
+            return Filter(session.GetPremiumMarketItems(), genderFilter, jobFilter, searchString).Cast<MarketItem>().ToList();
+        }
+        if (!TableMetadata.MeretMarketCategoryTable.Entries.TryGetValue((int) section, out IReadOnlyDictionary<int, MeretMarketCategoryTable.Tab>? sectionDictionary) ||
+            !sectionDictionary.TryGetValue(tabId, out MeretMarketCategoryTable.Tab? tab)) {
+            return new List<MarketItem>();
+        }
+
+        // get any sub tabs
+        int[] tabIds = new[] {tabId}.Concat(tab.SubTabIds).ToArray();
+
+        ICollection<PremiumMarketItem> results = session.GetPremiumMarketItems(tabIds);
+        results = Filter(results, genderFilter, jobFilter, searchString);
+        return Sort(results, sortBy, tab.SortGender, tab.SortJob);
+    }
+
+    private static StringCode Pay(GameSession session, long price, MeretMarketCurrencyType currencyType) {
+        switch (currencyType) {
+            case MeretMarketCurrencyType.Meso:
+                if (session.Currency.CanAddMeso(-price) != -price) {
+                    return StringCode.s_err_lack_meso;
+                }
+                session.Currency.Meso -= price;
+                return StringCode.s_empty_string;
+            case MeretMarketCurrencyType.Meret:
+                if (session.Currency.CanAddMeret(-price) != -price) {
+                    return StringCode.s_err_lack_merat;
+                }
+                session.Currency.Meret -= price;
+                return StringCode.s_empty_string;
+            case MeretMarketCurrencyType.RedMeret: // TODO: Implement Red Meret
+                return StringCode.s_err_lack_merat_red;
+            default:
+                return StringCode.s_err_lack_money;
+        }
+    }
+
+    private static ICollection<PremiumMarketItem> Filter(IEnumerable<PremiumMarketItem> items, GenderFilterFlag gender, JobFilterFlag job, string searchString) {
```

Comment:
I would return `IEnumerable` here so you don't have to allocate the `List`. Since these intermediate steps are chained and you only need the list at the very end.